rat/compiler.py

```diff
@@ -498,207 +577,155 @@ def _prepare_for_scans(self):
-                        carry_tuple_string = f"({','.join(carry_strings)})"
-                        next_carry_tuple_string = f"({','.join(['next_value'] + carry_strings[:-1])})"
+        for top_expr in list(reversed(self.expr_tree_list)):
+            if not isinstance(top_expr, ast.Assignment):
+                continue
 
-                        code += f"    def {scan_function_name}(carry, index):\n"
-                        if need_first_in_group_indicator:
-                            code += f"        carry = jax.lax.cond(first_in_group_indicators['{lhs_key}'][index], lambda : {zero_tuple_string}, lambda : carry)\n"
-                        code += f"        {carry_tuple_string} = carry\n"
-                        code += f"        next_value = {top_expr.rhs.code(scalar = True)}\n"
-                        code += f"        return {next_carry_tuple_string}, next_value\n"
+            codegen_visitor = codegen_backends.TransformedParametersVisitor(
+                self.symbol_table, self._get_primary_symbol_from_statement(top_expr), indent=4
+            )
 
-                        code += f"\n    _, parameters['{lhs_key}'] = jax.lax.scan({scan_function_name}, {zero_tuple_string}, jax.numpy.arange({lhs_size}))\n"
-                    else:
-                        code += f"    parameters['{lhs_key}'] = {top_expr.rhs.code()}\n"
-        code += "\n    return parameters\n"
+            top_expr.accept(codegen_visitor)
+            self.generated_code += codegen_visitor.get_expression_string()
+            self.generated_code += "\n"
 
-        # Generate code for evaluating densities
-        code += f"\ndef evaluate_densities(data, subscripts, parameters):\n" f"    target = 0.0\n"
+        self.generated_code += "\n"
+        self.generated_code += "    return parameters\n\n"
 
-        for top_expr in ordered_expr_trees:
-            if isinstance(top_expr, ops.Distr):
-                code += f"    target += jax.numpy.sum({top_expr.code()})\n"
-        code += "\n    return target"
+    def codegen_evaluate_densities(self):
+        self.generated_code += "def evaluate_densities(data, subscripts, parameters):\n"
+        self.generated_code += "    target = 0.0\n"
 
-        return code
+        for top_expr in self.expr_tree_list:
+            if not isinstance(top_expr, ast.Distr):
+                continue
 
-    def compile(
-        self,
-    ) -> Tuple[
-        Dict[str, variables.Data],
-        Dict[str, variables.Param],
-        Dict[str, variables.AssignedParam],
-        Dict[str, variables.SubscriptUse],
-        Dict[str, numpy.ndarray],
-        str,
-    ]:
-        if self.parameter_variables:
-            raise Exception("Compiler.compile() may be invoked only once per instance. Create a new instance to recompile.")
+            codegen_visitor = codegen_backends.EvaluateDensityVisitor(self.symbol_table, self._get_primary_symbol_from_statement(top_expr))
+            top_expr.accept(codegen_visitor)
```

Comment:
I want to replace all the visitor stuff with:

```python
def evaluate_density_visitor(base_expr):
  for expr in base_expr:
    match expr:
      case ast.Subscript():
        ...
      case ast.Param():
        ...
```

If we do this we can delete all the `accept` functions from the ast and there's less function boilerplate on the codegen_backend part.

Reply:
The reason for this is because different backends can have different behavior for some nodes.
For example, if a backend subclass is identical to the evaluate density visitor except for a single node, we'd have to rewrite all the match cases.

Reply:
Mmm I see. I do not like the visitor pattern, Sam I Am. Hmmm hmmmm hmmmm, is there anyway around this? Otherwise I am going to think about how to attack it from a do-we-need-it-perspective lol.

Reply:
If there's no way around this and we need it then I'll be convinced lol, maybe I just gotta stare at it more.

Reply:
> if a backend subclass is identical to the evaluate density visitor except for a single node, we'd have to rewrite all the match cases.

Okay so staring at this code we have EvaluateDensityVisitor and TransformedparametersVisitor. In each of these cases, the member functions have special overrides for certain things, like `visit_Param`, and `visit_Data`.

As we're parsing an expression tree, because of how the accept/visit logic works, if we hit a Param or a Data, we'll call our own special stuff before passing off to the more generic functions up the line (in BaseVisitor).

What if instead we did something like this:

```python
class EvaluateDensityVisitor(BaseVisitor):
  def my_data_logic(self, expr):
    pass

  def my_param_logic(self, expr):
    pass

  def generate(base_expr):
    for expr in base_expr:
      match expr:
        case ast.Data():
          self.my_data_logic(expr)
        case ast.Param():
          self.my_param_logic(expr)
        case _:
          self.generic_logic(expr) # This is implemented in BaseVisitor
```

Reply:
Shit, I guess generic logic needs to somehow call back here again?

Reply:
Noooooooo

Reply:
We made this change